openvdb/points/AttributeArrayString.cc

```diff
@@ -20,86 +20,144 @@ namespace points {
-    }
-
-    Name getStringKey(const StringIndexType index)
+    Name getStringKey(const Index index)
     {
         return "string:" + std::to_string(index - 1);
     }
 
-    StringIndexType getStringIndex(const Name& key)
-    {
-        const Name indexStr = key.substr(7, key.size() - 7);
-        // extract the index as an unsigned integer
-        return static_cast<StringIndexType>(std::stoul(indexStr)) + 1;
-    }
-
 } // namespace
 
 
 ////////////////////////////////////////
 
 
+// StringMetaCache implementation
+
+
+StringMetaCache::StringMetaCache(const MetaMap& metadata)
+{
+    this->reset(metadata);
+}
+
+
+void StringMetaCache::insert(const Name& key, Index index)
+{
+    mCache[key] = index;
+}
+
+
+void StringMetaCache::reset(const MetaMap& metadata)
+{
+    mCache.clear();
+
+    // populate the cache
+
+    for (auto it = metadata.beginMeta(), itEnd = metadata.endMeta(); it != itEnd; ++it) {
+        const Name& key = it->first;
+        const Metadata::Ptr& meta = it->second;
+
+        // attempt to cast metadata to StringMetadata
+        const StringMetadata* stringMeta = dynamic_cast<StringMetadata*>(meta.get());
+        if (!stringMeta)                            continue;
+
+        // string attribute metadata must have a key that starts "string:"
+        if (key.compare(0, 7, "string:") != 0)      continue;
+
+        // remove "string:" and cast to Index
+        Index index = 1 + static_cast<Index>(
+            std::stoul(key.substr(7, key.size() - 7)));
+
+        // add to the cache
+        this->insert(stringMeta->value(), index);
+    }
+}
+
+
+////////////////////////////////////////
+
 // StringMetaInserter implementation
 
 
 StringMetaInserter::StringMetaInserter(MetaMap& metadata)
     : mMetadata(metadata)
     , mIdBlocks()
-    , mValues()
+    , mCache()
 {
     // populate the cache
     resetCache();
 }
 
 
-void StringMetaInserter::insert(const Name& name)
+bool StringMetaInserter::hasKey(const Name& key) const
+{
+    return mCache.map().find(key) != mCache.map().end();
+}
+
+
+bool StringMetaInserter::hasIndex(Index index) const
 {
-    using IterT = std::vector<std::pair<Index, Index>>::iterator;
+    return bool(mMetadata[getStringKey(index)]);
+}
+
 
-    // if name already exists, do nothing
+Index StringMetaInserter::insert(const Name& name, Index hint)
+{
+    using IterT = IndexPairArray::iterator;
 
-    if (mValues.count(name))  return;
+    // if name already exists, return the index
 
-    // look through the id blocks for the next available index
+    const auto& cacheMap = mCache.map();
+    auto it = cacheMap.find(name);
+    if (it != cacheMap.end()) {
+        return it->second;
+    }
 
     Index index = 1;
+
+    Name hintKey;
+    bool canUseHint = false;
+
+    // hint must be non-zero to have been requested
+
+    if (hint > Index(0)) {
+        hintKey = getStringKey(hint);
+        // check if hint is already in use
+        if (!bool(mMetadata[hintKey])) {
```

Comment:
As this requires a map lookup is the usage of hint really more efficient?

Reply:
It's not so much for efficiency here, it's more that you can insert using non-sequential indices so that you don't need to move any AttributeArray data if you can align the indices which _is_ much more efficient. You can do it without this method by creating a temporary data structure yourself, but it's a convenience to be able to hint at a preferred index.

Reply:
Right, so as i understand it, you shouldn't then need to run the loop which checks for the next valid index? i.e., instead of

```
    IterT iter = mIdBlocks.begin();
    for (; iter != mIdBlocks.end(); ++iter) {
        const Index start = iter->first;
        const Index end = start + iter->second;

        if (index < start || index >= end) break;
        if (!canUseHint)    index = end;
    }
```
it's
```
if (!canUseHint) {
    IterT iter = mIdBlocks.begin();
    for (; iter != mIdBlocks.end(); ++iter) {
        const Index start = iter->first;
        const Index end = start + iter->second;

        if (index < start || index >= end) break;
        index = end;
    }
}
```